lib/spack/spack/test/url_substitution.py

```diff
@@ -25,61 +25,50 @@
+     'https://github.com/losalamos/CLAMR/blob/packages/PowerParser_v2.0.7.tgz?raw=true'),
+    ('https://github.com/losalamos/CLAMR/blob/packages/PowerParser_v2.0.7.tgz?raw=true', '4.7',
+     'https://github.com/losalamos/CLAMR/blob/packages/PowerParser_v4.7.tgz?raw=true'),
+    # Package name contains regex characters
+    ('http://math.lbl.gov/voro++/download/dir/voro++-0.4.6.tar.gz', '1.2.3',
+     'http://math.lbl.gov/voro++/download/dir/voro++-1.2.3.tar.gz'),
+])
+def test_url(request):
+    """Returns a tuple containing the base url, the version to be
+    substituted and the expected result.
+    """
+    return request.param
 
-    def test_different_version(self):
-        # Test a completely different version syntax
-        self.check(
-            'http://www.prevanders.net/libdwarf-20130729.tar.gz', '8.12',
-            'http://www.prevanders.net/libdwarf-8.12.tar.gz')
 
-    def test_double_version(self):
-        # Test a URL where the version appears twice
-        # It should get substituted both times
-        self.check(
-            'https://github.com/hpc/mpileaks/releases/download/v1.0/mpileaks-1.0.tar.gz', '2.1.3',
-            'https://github.com/hpc/mpileaks/releases/download/v2.1.3/mpileaks-2.1.3.tar.gz')
+def test_url_substitution(test_url):
 
-    def test_partial_version_prefix(self):
-        # Test now with a partial prefix earlier in the URL
-        # This is hard to figure out so Spack only substitutes
-        # the last instance of the version
-        self.check(
-            'https://www.open-mpi.org/software/ompi/v2.1/downloads/openmpi-2.1.0.tar.bz2', '2.2.0',
-            'https://www.open-mpi.org/software/ompi/v2.1/downloads/openmpi-2.2.0.tar.bz2')
-        self.check(
-            'https://www.open-mpi.org/software/ompi/v2.1/downloads/openmpi-2.1.0.tar.bz2', '2.2',
-            'https://www.open-mpi.org/software/ompi/v2.1/downloads/openmpi-2.2.tar.bz2')
-
-    def test_no_separator(self):
-        # No separator between the name and version of the package
-        self.check(
-            'file://{0}/turbolinux702.tar.gz'.format(os.getcwd()), '703',
-            'file://{0}/turbolinux703.tar.gz'.format(os.getcwd()))
-
-    def test_github_raw(self):
-        self.check(
-            'https://github.com/losalamos/CLAMR/blob/packages/PowerParser_v2.0.7.tgz?raw=true', '2.0.7',
-            'https://github.com/losalamos/CLAMR/blob/packages/PowerParser_v2.0.7.tgz?raw=true')
-        self.check(
-            'https://github.com/losalamos/CLAMR/blob/packages/PowerParser_v2.0.7.tgz?raw=true', '4.7',
-            'https://github.com/losalamos/CLAMR/blob/packages/PowerParser_v4.7.tgz?raw=true')
-
-    def test_regex(self):
-        # Package name contains regex characters
-        self.check(
-            'http://math.lbl.gov/voro++/download/dir/voro++-0.4.6.tar.gz', '1.2.3',
-            'http://math.lbl.gov/voro++/download/dir/voro++-1.2.3.tar.gz')
+    base_url, version, expected = test_url
```

Comment:
I used it to simplify the fetch tests a while ago -- it's sort of a tradeoff because you can't reuse the parametrization like you can with a fixture.  But it makes sense here because you can parameterize multiple arguments.  OT: somewhat annoying that they used a less common spelling of "parameterize".